docs/getting-started/rulesets.md

```diff
@@ -199,3 +198,141 @@ rules:
+To give you an example of a good code
+```js
+module.exports = (obj) => {
+  for (const [key, value] of Object.entries(obj)) {
+    // this is a perfectly fine code
+  }
+};
+```
+You do not need to provide any shim for `Object.entries` or use [regenerator](https://facebook.github.io/regenerator/) for `for of` loop.
+As stated, ou cannot use ES Modules, so the following code is considered sa invalid and won't work correctly.
+```js
+export default (obj) => {
+  for (const [key, value] of Object.entries(obj)) {
+    // this is a perfectly fine code
+  }
+};
+```
+The following code won't work properly either
+ ```js
+const foo = require('./foo'); // require is not available (see note below)
+
+module.exports = (obj) => {
+  for (const [key, value] of Object.entries(obj)) {
+    // this is a perfectly fine code
+  }
+};
+```
+
+If you have any module system, you need to use some bundler, preferably Rollup.js as it generates efficient bundles.
+We are still evaluating the idea of supporting ESModule and perhaps we decide to bring support for ES Modules at some point,
+yet for now you cannot use them.
+
+#### How do I create a function?
+
+The process of creating a function involves 2 steps:
+
+- create a js file inside of a directory called `functions` that should be placed next to your ruleset file
+- create a `functions` array in your ruleset if you haven't done it yet and place a string with the filename without `.js` extension
+
+Example:
+
+`functions/abc.js`
+
+```js
+module.exports = (targetVal) => {
+  if (targetVal !== 'abc') {
+    return [
+      {
+        message: 'Value must equal "abc".',
+      },
+    ];
+  }
+};
+```
+
+`my-ruleset.yaml`
+
+```yaml
+functions: [abc]
+rules:
+  my-rule:
+    message: "{{error}}"
+    given: "$.info"
+    then:
+      function: "abc"
+```
+
+Optionally, if you'd like to validate the data that it passed to abc function before the function gets executed, you can provide a JSON Schema.
+You can do it as follows
+
+```yaml
+functions: [[abc, { type: "string" }]] # can be any valid JSONSchema7
+rules:
+  my-rule:
+    message: "{{error}}"
+    given: "$.info"
+    then:
+      function: "abc"
+```
+
+If for some reason, you do not want to place your functions in a directory called `functions`, you can specify a custom path.
+
+```yaml
+functionsDir: "./my-functions" # any relative path to the ruleset is okay
+functions: [[abc, { type: "string" }]] # can be any valid JSONSchema7
+rules:
+  my-rule:
+    message: "{{error}}"
+    given: "$.info"
+    then:
+      function: "abc"
+```
+
+Spectral would look for functions in a `my-functions` directory now.
+
+#### Inheritance
+
+Core functions can be overridden with custom rulesets, so if you'd like to make your own truthy go ahead. 
+Other than that, custom functions are only available in the ruleset which defines them, so loading a foo in one ruleset will not clobber a foo in another ruleset.
```

Comment:
🤔 what does this mean? If somebody creates a ruleset that has some snazzy custom functions, I cannot extend that ruleset and add some rules of my own that use those functions?

Reply:
You can (see tests). Note, this is not recommended - if you extend a few rulesets that implement the same method and change the order of those rulesets, a different function will be picked up.

Reply:
https://github.com/stoplightio/spectral/blob/2defef04b140d3e93179345315002b458936b3d5/src/rulesets/mergers/__tests__/functions.jest.test.ts
https://github.com/stoplightio/spectral/blob/2defef04b140d3e93179345315002b458936b3d5/src/__tests__/linter.jest.test.ts#L27#L76 
https://github.com/stoplightio/spectral/blob/2defef04b140d3e93179345315002b458936b3d5/src/rulesets/__tests__/reader.jest.test.ts#L357#L412

Reply:
@P0lip If you extend two rulesets which declare an `apples` function, and you try to use the apples function in your ruleset, which one do you get?

@marbemac same scenario but what would you expect?

Reply:
> @P0lip If you extend two rulesets which declare an apples function, and you try to use the apples function in your ruleset, which one do you get?

@philsturgeon the most recent one one. These 2 extended rulesets would have to its own `apples` function though.